subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/compile/JavaCompileCompatibilityIntegrationTest.groovy

```diff
@@ -0,0 +1,587 @@
+                doLast {
+                    logger.lifecycle("project.sourceCompatibility = '\${projectSourceCompat}'")
+                    logger.lifecycle("project.targetCompatibility = '\${projectTargetCompat}'")
+                    logger.lifecycle("task.sourceCompatibility = '\$sourceCompatibility'")
+                    logger.lifecycle("task.targetCompatibility = '\$targetCompatibility'")
+                }
+            }
+        """
+
+        when:
+        withInstallations(jdk11).run(":compileJava")
+
+        then:
+        outputContains("project.sourceCompatibility = '11'")
+        outputContains("project.targetCompatibility = '11'")
+        outputContains("task.sourceCompatibility = '$sourceOut'")
+        outputContains("task.targetCompatibility = '$targetOut'")
+        classJavaVersion(javaClassFile("Foo.class")) == JavaVersion.toVersion(targetOut)
+
+        where:
+        source | target | sourceOut | targetOut
+        '9'    | '10'   | '9'       | '10'
+        '9'    | 'none' | '9'       | '9'
+        'none' | 'none' | '11'      | '11'
+    }
+
+    def "configuring toolchain on java extension with source and target compatibility is supported"() {
+        def jdk = Jvm.current()
+        def prevJavaVersion = JavaVersion.toVersion(jdk.javaVersion.majorVersion.toInteger() - 1)
+        buildFile << """
+            apply plugin: 'java'
+
+            java {
+                sourceCompatibility = JavaVersion.toVersion('$prevJavaVersion')
+                targetCompatibility = JavaVersion.toVersion('$prevJavaVersion')
+                toolchain {
+                    languageVersion = JavaLanguageVersion.of(${jdk.javaVersion.majorVersion})
+                }
+            }
+
+            compileJava {
+                def projectSourceCompat = project.java.sourceCompatibility
+                def projectTargetCompat = project.java.targetCompatibility
+                doLast {
+                    logger.lifecycle("project.sourceCompatibility = '\${projectSourceCompat}'")
+                    logger.lifecycle("project.targetCompatibility = '\${projectTargetCompat}'")
+                    logger.lifecycle("task.sourceCompatibility = '\$sourceCompatibility'")
+                    logger.lifecycle("task.targetCompatibility = '\$targetCompatibility'")
+                }
+            }
+        """
+
+        when:
+        withInstallations(jdk).run(":compileJava")
+
+        then:
+        outputContains("project.sourceCompatibility = '$prevJavaVersion'")
+        outputContains("project.targetCompatibility = '$prevJavaVersion'")
+        outputContains("task.sourceCompatibility = '$prevJavaVersion'")
+        outputContains("task.targetCompatibility = '$prevJavaVersion'")
+        classJavaVersion(javaClassFile("Foo.class")) == JavaVersion.toVersion(prevJavaVersion)
+    }
+
+    def "configuring toolchain on java extension and clearing source and target compatibility is supported"() {
```

Comment:
Question: This seems like an odd test.  I wouldn't normally test for strange user behavior like this in an integration test. This seems like the sort of thing that might be in a unit test.  Why would we choose to test this?  Is this in response to a defect?

Reply:
I agree. Previously, it was a test that checked ability to clear values, do to toolchains being incompatible (failing validation) with source compatibility on the `java` extension level. The test was adopted with minimal changes to make it work.

I also agree, that it does not make a lot of sense anymore. I've removed it now.